OpenNefia.Content/CharaMake/CharaMakeRaceClassLayer.cs

```diff
@@ -0,0 +1,386 @@
+using OpenNefia.Core.Prototypes;
+using OpenNefia.Core.Rendering;
+using OpenNefia.Core.UI.Element;
+using OpenNefia.Core.Utility;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OpenNefia.Content.CharaMake
+{
+    [Localize("Elona.CharaMake.RaceSelect")]
+    public class CharaMakeRaceSelectLayer : CharaMakeRaceClassLayer
+    {
+        private const string ResultName = "race";
+        public override IEnumerable<RaceClass> GetData()
+        {
+            return _prototypeManager.EnumeratePrototypes<RacePrototype>().Select(x => new RaceClass(x));
+        }
+
+        protected override void Select(RaceClass item)
+        {
+            Finish(new CharaMakeResult(new Dictionary<string, object>
+            {
+                { ResultName, item.Data }
+            }));
+        }
+    }
+    [Localize("Elona.CharaMake.ClassSelect")]
+    public class CharaMakeClassSelectLayer : CharaMakeRaceClassLayer
+    {
+        private const string ResultName = "class";
+        public override IEnumerable<RaceClass> GetData()
+        {
+            return _prototypeManager.EnumeratePrototypes<ClassPrototype>().Select(x => new RaceClass(x));
+        }
+
+        protected override void Select(RaceClass item)
+        {
+            Finish(new CharaMakeResult(new Dictionary<string, object>
+            {
+                { ResultName, item.Data }
+            }));
+        }
+    }
+
+    public abstract class CharaMakeRaceClassLayer : CharaMakeLayer
+    {
+        public class RaceClass
+        {
+            public IPrototype Data;
+
+            public RaceClass(IPrototype data)
+            {
+                Data = data;
+            }
+            public string GetString(string suffix = "")
+            {
+                switch (Data)
+                {
+                    case RacePrototype race:
+                        return Loc.GetPrototypeString<RacePrototype>(new(race.ID), suffix)!;
+                    case ClassPrototype @class:
+                        return Loc.GetPrototypeString<ClassPrototype>(new(@class.ID), suffix)!;
+                    default:
+                        return string.Empty;
+                }
+            }
+
+            public Dictionary<PrototypeId<SkillPrototype>, int> GetSkills()
+            {
+                switch (Data)
+                {
+                    case RacePrototype race:
+                        return race.BaseSkills;
+                    case ClassPrototype @class:
+                        return @class.BaseSkills;
+                    default:
+                        return new();
+                }
+            }
+        }
+        public class RaceClassCell : UiListCell<RaceClass>
+        {
+            
+            public RaceClassCell(RaceClass wrapper) 
+                : base(wrapper, new UiText(UiFonts.ListText))
+            {
+                Text = Data.GetString("Name");
```

Comment:
In vanilla, the text `(extra)` gets prepended if this class/race is an "extra" class (not normally selectable unless you set a config option). Basically it would be the same thing as getting the skills, since class/race prototypes (should) have an `IsExtra` boolean property.

Reply:
done in dd52918